### PR TITLE
DEV-8329 Generate index of notebooks based on docstring metadata

### PR DESCRIPTION
**Pull request by rizwansaeed**
Auto generate an index of notebooks and their features by adding a docstring at the start of each notebook.

By convention where the first cell is code cell containing a docstring of the form:

```
"""Valuation with recipes

This notebooks shows how to value a portfolio using recipes with different pricing sources

Attributes
----------
valuation
transactions
recipes
manifests
"""
```

it will get indexed into a markdown file, see /examples/README.md for sample output

TODO:
* ~get review comments~
* ~update remaining notebooks~
* add to travis build

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -8,3 +8,6 @@ numpy
 requests
 lusidtools
 lusid-jam
+docstring-parser
+parameterized
+chevron
```
Comment (MikeMcGarry) [inline requirements.txt:13]: Can we please add these to a requirments.dev.txt file instead as they are not required to run the notebooks.

File: docgen/nbmeta.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class NbMeta:
+
```
Comment (MikeMcGarry) [inline docgen/nbmeta.py:2]: What is this class? Can we please add a Docstring?

File: docgen/parser.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import itertools
+import nbformat
+import os
+
+from pathlib import Path
+
+from docgen import NbMeta
+
+
+def find_nbs(nb_root):
```
Comment (MikeMcGarry) [inline docgen/parser.py:12]: Can we please add a Docstring?

File: docgen/parser.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    for root, dirs, files in os.walk(nb_root, topdown=False):
+        for name in files:
+            if name.lower().endswith(".ipynb"):
+                yield os.path.join(root, name)
+
+
+def sanitize_docstring(raw_str):
```
Comment (MikeMcGarry) [inline docgen/parser.py:19]: Can we please add a Docstring?

File: docgen/parser.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    stripped_str = raw_str.strip()
+    if stripped_str.startswith('"""') and stripped_str.endswith('"""'):
+        return stripped_str[3:-3]
+
+
+def nb_relative_path(abs_path):
```
Comment (MikeMcGarry) [inline docgen/parser.py:25]: Can we please add a Docstring?

File: docgen/parser.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    rel_path = os.path.relpath(abs_path)
+    return rel_path[3:]
+
+
+def process_nb(nb_path):
```
Comment (MikeMcGarry) [inline docgen/parser.py:30]: Can we please add a Docstring?

File: docgen/parser.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    return NbMeta(
+        nb_relative_path(os.path.dirname(nb_path)),
+        os.path.basename(nb_path),
+        doc_str_obj.short_description,
+        doc_str_obj.long_description,
+        sorted([m.arg_name for m in doc_str_obj.meta])
+    )
+
+
+def parse(nb_root):
```
Comment (MikeMcGarry) [inline docgen/parser.py:51]: Can we please add a Docstring?

File: docgen/parser.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    return [
+        meta for meta in
+        [process_nb(nb) for nb in find_nbs(nb_root)]
+        if meta is not None
+    ]
+
+
+def build_doc(meta, template):
```
Comment (MikeMcGarry) [inline docgen/parser.py:59]: Can we please add a Docstring?

File: docgen/parser.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    nbs = [{"k": key, "v": sorted([*values], key=lambda m: m.filename)} for key, values in itertools.groupby(meta, lambda m: m.path)]
+
+    # sort by relative path
+    nbs.sort(key=lambda n: n["k"])
+
+    with open(template, 'r') as f:
+        return chevron.render(f, {'paths': nbs})
+
+
+def save_index_page(path, doc):
```
Comment (MikeMcGarry) [inline docgen/parser.py:74]: Can we please add a Docstring?

File: docgen/parser.py (hunk 3 above)
Comment (rizwansaeed) [reply docgen/parser.py:12]: done

File: docgen/parser.py (hunk 4 above)
Comment (rizwansaeed) [reply docgen/parser.py:19]: done

File: docgen/parser.py (hunk 5 above)
Comment (rizwansaeed) [reply docgen/parser.py:25]: done

File: docgen/parser.py (hunk 6 above)
Comment (rizwansaeed) [reply docgen/parser.py:30]: done

File: docgen/parser.py (hunk 7 above)
Comment (rizwansaeed) [reply docgen/parser.py:51]: done

File: docgen/parser.py (hunk 8 above)
Comment (rizwansaeed) [reply docgen/parser.py:59]: done

File: docgen/parser.py (hunk 9 above)
Comment (rizwansaeed) [reply docgen/parser.py:74]: done

**Comment by rizwansaeed** [comment]
Updated with review feedback, thanks for keeping me honest!

File: requirements.txt [hunk 10]
```diff
@@ -8,3 +8,5 @@ numpy
 requests
 lusidtools
 lusid-jam
+docstring-parser
+chevron
```
Comment (MikeMcGarry) [inline requirements.txt:12]: Are these dev requirements?

File: examples/README.md [hunk 11]
```diff
@@ -1,6 +1,22 @@
...
+## examples/use-cases/ibor
+| filename | title | description | features |
+| --- | --- | --- | --- |
+| Generating Corporate actions natively in LUSID.ipynb | Corporate Actions in LUSID |  | corporate actions, derived portfolios, holdings, transactions |
+| Valuations with recipes.ipynb | Valuation with recipes | This notebooks shows how to value a portfolio using recipes with different pricing sources | manifests, recipes, transactions, valuation |
+
+
+| :warning: This file is generated, any direct edits may be lost |
+| --- |
+
```
Comment (MikeMcGarry) [inline examples/README.md:22]: How come this doesn't have all the notebooks?

File: requirements.txt (hunk 10 above)
Comment (rizwansaeed) [reply requirements.txt:12]: they're required to generate the documentation, I think they should stay in main requirement.txt

File: examples/README.md (hunk 11 above)
Comment (rizwansaeed) [reply examples/README.md:22]: I've only updated a couple to check its working, see the TODO above

**Pull request merged**